### PR TITLE
Created TrackingInterceptor as an alternative to TrackedDioClient

### PR DESCRIPTION
**Pull request by lohnn**
The TrackedDioClient had some shortcomings, so I created an interceptor for Dio to have flexibility and support Dio clients without having to keep implementation up to date with DioMixin interface, as Interceptors will most likely have a more stable API surface.

Current shortcomings of this implementation is that StackTrace in error flow will not work until PR https://github.com/flutterchina/dio/pull/1503 is merged.

### CODE REVIEW

**Comment by teodor-appd** [comment]
What's the status on this one, @lohnn?

**Comment by lohnn** [comment]
We are still waiting for the Dio team to become active again and merge my PR on their side. Once that is done, this PR should be all good to go.
We mailed one of the maintainers in hopes of getting it going.

**Comment by lohnn** [comment]
In regards to your Discord account, it says you are "Deleted user", not sure if it's possible to get it back tho. 🤔

**Comment by spydon** [comment]
How's this PR looking now when Dio 5.0.0 is out?
It's a bit unclear what happened to all the PRs when they first moved to Diox and then back to Dio.

**Comment by teodor-appd** [comment]
> How's this PR looking now when Dio 5.0.0 is out? It's a bit unclear what happened to all the PRs when they first moved to Diox and then back to Dio.

We've added supporting Dio 5.0.0 to the roadmap. However, it will be scheduled for next month. Regarding the PRs, I don't know if you're talking about this project's or Dio's repo. You might be talking about the latter, where I'm unfamiliar with the context, but maybe @lohnn can help.

**Comment by lohnn** [comment]
I'll take a look at if I need to reopen the https://github.com/flutterchina/dio/pull/1503 pr or not for this one to work. Just haven't had time to do so yet.

**Comment by lohnn** [comment]
Okay, so I've done a small test locally now and the status of the connected PR https://github.com/flutterchina/dio/pull/1503 is that I'll have to reopen that one, fix any potential merge conflicts or breaking changes and still wait for approval.
Good thing is that it's a new team of people and it will most likely be easier to get it merged.

**Comment by lohnn** [comment]
As the old PR at Dio was closed, there is a new one https://github.com/cfug/dio/pull/1721, hopefully this one is merged soon, then this PR should be able to be merged once they release a new Dio version.

**Comment by AlexV525** [comment]
> How's this PR looking now when Dio 5.0.0 is out? It's a bit unclear what happened to all the PRs when they first moved to Diox and then back to Dio.

I just realized that we've never talked about this. From the start, we assume Dio will not get maintained anymore, so we've picked valid (at the first few looks) into our hard fork. Then we got the ownership, so those merged PRs were merged back to the original repo, again. If you can see a linked PR targeting Diox that is merged, it's merged in the Dio too.

**Comment by lohnn** [comment]
Thanks to https://github.com/cfug/dio/pull/1722 this is no longer blocked.
The failing tests are in classes that I have, as far as I can see, not touched. Not sure what to do with those.

**Comment by lohnn** [comment]
I'll have a look at your changes and see, and let's then try to get this merged shortly 👍

File: lib/src/tracked_clients/tracked_dio_interceptor.dart [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  @override
+  void onResponse(
+    Response<dynamic> response,
+    ResponseInterceptorHandler handler,
+  ) async {
+    try {
+      final tracker =
+          _activeTrackers.remove(response.requestOptions.extra[_trackerId]);
+      if (tracker != null) {
+        await tracker.setResponseStatusCode(response.statusCode ?? 404);
```
Comment (lohnn) [inline lib/src/tracked_clients/tracked_dio_interceptor.dart:61]: The fallback should in theory not be needed, as a successful response should always have a status code.
On the other hand, sure, let's keep the fallback if Dio would change their implementation.

**Pull request merged**